File: web/src/api/notify/templates.ts
import { http } from '@/lib/http';

import type { NotifyCategory } from './index';

export interface NotifyTemplate {
  id: string;
  organization_id?: string;
  name: string;
  body: string;
  format?: 'text' | 'markdown' | 'html';
  category: NotifyCategory;
  created_at_micros?: number;
  updated_at_micros?: number;
}

interface NotifyTemplateWire
  extends Omit<
    NotifyTemplate,
    'category' | 'created_at_micros' | 'updated_at_micros'
  > {
  category?: NotifyCategory;
  created_at?: number;
  updated_at?: number;
  created_at_micros?: number;
  updated_at_micros?: number;
}

function normalize(row: NotifyTemplateWire): NotifyTemplate {
  const {
    category,
    created_at,
    updated_at,
    created_at_micros,
    updated_at_micros,
    ...rest
  } = row;
  const out: NotifyTemplate = { ...rest, category: category ?? 'alert' };
  const created = created_at_micros ?? created_at;
  const updated = updated_at_micros ?? updated_at;
  if (created !== undefined) out.created_at_micros = created;
  if (updated !== undefined) out.updated_at_micros = updated;
  return out;
}

export async function list(): Promise<NotifyTemplate[]> {
  const { data } = await http.get<NotifyTemplateWire[]>('/notify/templates');
  return data.map(normalize);
}

export async function get(id: string): Promise<NotifyTemplate> {
  const { data } = await http.get<NotifyTemplateWire>(
    `/notify/templates/${encodeURIComponent(id)}`,
  );
  return normalize(data);
}

export type NotifyTemplateInput = Pick<
  NotifyTemplate,
  'name' | 'body' | 'format' | 'category'
>;

export async function create(payload: NotifyTemplateInput): Promise<NotifyTemplate> {
  const { data } = await http.post<NotifyTemplateWire>(
    '/notify/templates',
    payload,
  );
  return normalize(data);
}

export async function update(id: string, payload: NotifyTemplateInput): Promise<NotifyTemplate> {
  const { data } = await http.put<NotifyTemplateWire>(
    `/notify/templates/${encodeURIComponent(id)}`,
    payload,
  );
  return normalize(data);
}

export async function remove(id: string): Promise<void> {
  await http.delete(`/notify/templates/${encodeURIComponent(id)}`);
}

export type NotifyTemplateFieldGroup =
  | 'event'
  | 'message'
  | 'rule'
  | 'incident'
  | 'trigger'
  | 'labels'
  | 'annotations'
  | 'schedule'
  | 'oncall'
  | 'override';

export interface NotifyTemplateField {
  key: string;
  token: string;
  group: NotifyTemplateFieldGroup;
  categories: NotifyCategory[];
  example: string;
  event_types: string[];
}

export interface NotifyTemplatePreset {
  key: string;
  category: NotifyCategory;
  event_type: string;
  format: NonNullable<NotifyTemplate['format']>;
  body: string;
}

export interface NotifyTemplateFields {
  fields: NotifyTemplateField[];
  presets: NotifyTemplatePreset[];
  label_keys: string[];
  annotation_keys: string[];
}

export async function listFields(): Promise<NotifyTemplateFields> {
  const { data } = await http.get<NotifyTemplateFields>(
    '/notify/template-fields',
  );
  return data;
}

File: bin/molesignal/src/api/http/routes/notify/templates.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 MoleSignal Authors

//! Notify 模板 CRUD。

use std::collections::BTreeSet;

use axum::{
    Extension, Json, Router,
    extract::{Path, State},
    routing::get,
};
use serde::{Deserialize, Serialize};

use crate::{
    api::{
        AppState,
        http::federation::{delete_payload, emit_cud},
    },
    app::{iam::IamContext, notify::validate_notify_template_body},
    domain::{
        federation::{CudAction, ResourceKind},
        iam::permission,
        notify::{
            preference::NotifyCategory,
            template::{
                NotifyTemplateField, NotifyTemplatePreset, notify_template_field_catalog,
                notify_template_preset_catalog,
            },
        },
    },
    infra::persistence::repositories::notify::NotifyTemplateRecord,
    shared::{Error, Result, ids::Id, time::TimestampMicros},
};

pub(super) fn routes() -> Router<AppState> {
    Router::new()
        .route("/notify/templates", get(list).post(create))
        .route("/notify/template-fields", get(list_fields))
        .route(
            "/notify/templates/{id}",
            get(get_one).put(update).delete(delete),
        )
}

#[derive(Debug, Deserialize)]
struct WriteRequest {
    name: String,
    body: String,
    #[serde(default)]
    format: Option<String>,
    #[serde(default = "default_template_category")]
    category: NotifyCategory,
}

const ALLOWED_FORMATS: &[&str] = &["text", "markdown", "html"];

const fn default_template_category() -> NotifyCategory {
    NotifyCategory::Alert
}

fn validate_request(request: &WriteRequest) -> Result<String> {
    if request.name.trim().is_empty() {
        return Err(Error::invalid("name must not be empty"));
    }
    validate_notify_template_body(&request.body)?;
    let format = request.format.as_deref().unwrap_or("text");
    if !ALLOWED_FORMATS.contains(&format) {
        return Err(Error::invalid(
            "format must be one of: text | markdown | html",
        ));
    }
    Ok(format.to_string())
}

#[derive(Debug, Serialize)]
struct TemplateFieldsResponse {
    fields: Vec<NotifyTemplateField>,
    presets: Vec<NotifyTemplatePreset>,
    label_keys: Vec<String>,
    annotation_keys: Vec<String>,
}

#[permission("alerts.read")]
async fn list_fields(
    State(state): State<AppState>,
    Extension(context): Extension<IamContext>,
) -> Result<Json<TemplateFieldsResponse>> {
    let mut label_keys = BTreeSet::new();
    let mut annotation_keys = BTreeSet::new();
    if let Ok(rules) = state.alerting.service.list_rules(&context.org_id).await {
        for rule in rules {
            label_keys.extend(rule.labels.keys().cloned());
            annotation_keys.extend(rule.annotations.keys().cloned());
        }
    }
    if let Ok(incidents) = state
        .alerting
        .service
        .list_incidents_active(&context.org_id)
        .await
    {
        for incident in incidents {
            label_keys.extend(incident.labels.keys().cloned());
            annotation_keys.extend(incident.annotations.keys().cloned());
        }
    }
    Ok(Json(TemplateFieldsResponse {
        fields: notify_template_field_catalog(),
        presets: notify_template_preset_catalog(),
        label_keys: label_keys.into_iter().collect(),
        annotation_keys: annotation_keys.into_iter().collect(),
    }))
}

#[permission("alerts.read")]
async fn list(
    State(state): State<AppState>,
    Extension(context): Extension<IamContext>,
) -> Result<Json<Vec<NotifyTemplateRecord>>> {
    Ok(Json(state.alerting.templates.list(&context.org_id).await?))
}

#[permission("alerts.manage")]
async fn create(
    State(state): State<AppState>,
    Extension(context): Extension<IamContext>,
    Json(request): Json<WriteRequest>,
) -> Result<Json<NotifyTemplateRecord>> {
    let format = validate_request(&request)?;
    let existing = state.alerting.templates.list(&context.org_id).await?;
    if existing
        .iter()
        .any(|template| template.name == request.name)
    {
        return Err(Error::conflict("notify template name already exists"));
    }
    let now = TimestampMicros::now();
    let saved = state
        .alerting
        .templates
        .create(NotifyTemplateRecord {
            id: Id::new(),
            organization_id: context.org_id.clone(),
            name: request.name,
            body: request.body,
            format,
            category: request.category,
            created_at: now,
            updated_at: now,
        })
        .await?;
    emit_cud(
        &state,
        &context.org_id,
        ResourceKind::NotifyTemplate,
        CudAction::Created,
        &saved.id.0,
        &saved,
    )
    .await;
    Ok(Json(saved))
}

#[permission("alerts.read")]
async fn get_one(
    State(state): State<AppState>,
    Extension(context): Extension<IamContext>,
    Path(id): Path<String>,
) -> Result<Json<NotifyTemplateRecord>> {
    Ok(Json(
        state
            .alerting
            .templates
            .get(&context.org_id, &Id::from_string(id))
            .await?,
    ))
}

#[permission("alerts.manage")]
async fn update(
    State(state): State<AppState>,
    Extension(context): Extension<IamContext>,
    Path(id): Path<String>,
    Json(request): Json<WriteRequest>,
) -> Result<Json<NotifyTemplateRecord>> {
    let format = validate_request(&request)?;
    let existing = state
        .alerting
        .templates
        .get(&context.org_id, &Id::from_string(id))
        .await?;
    if state
        .alerting
        .templates
        .list(&context.org_id)
        .await?
        .iter()
        .any(|template| template.id != existing.id && template.name == request.name)
    {
        return Err(Error::conflict("notify template name already exists"));
    }
    let saved = state
        .alerting
        .templates
        .update(NotifyTemplateRecord {
            id: existing.id,
            organization_id: context.org_id.clone(),
            name: request.name,
            body: request.body,
            format,
            category: request.category,
            created_at: existing.created_at,
            updated_at: TimestampMicros::now(),
        })
        .await?;
    emit_cud(
        &state,
        &context.org_id,
        ResourceKind::NotifyTemplate,
        CudAction::Updated,
        &saved.id.0,
        &saved,
    )
    .await;
    Ok(Json(saved))
}

#[permission("alerts.manage")]
async fn delete(
    State(state): State<AppState>,
    Extension(context): Extension<IamContext>,
    Path(id): Path<String>,
) -> Result<Json<serde_json::Value>> {
    let id = Id::from_string(id);
    if state
        .alerting
        .notify_engine
        .list_policies(&context.org_id)
        .await?
        .iter()
        .any(|policy| policy.template_id.as_ref() == Some(&id))
    {
        return Err(Error::conflict(
            "notify template is referenced by a notify policy",
        ));
    }
    state
        .alerting
        .templates
        .delete(&context.org_id, &id)
        .await?;
    emit_cud(
        &state,
        &context.org_id,
        ResourceKind::NotifyTemplate,
        CudAction::Deleted,
        &id.0,
        &delete_payload(&id.0),
    )
    .await;
    Ok(Json(serde_json::json!({"deleted": true})))
}
